### PR TITLE
Limit zkApp Commands Hardcap

### PR DESCRIPTION
**Pull request by ejMina226**
Limit zkApp Commands to n >=0 in case of emergency
(See https://github.com/MinaProtocol/mina/pull/14783)

### CODE REVIEW

**Comment by ejMina226** [comment]
!ci-build-me

File: src/lib/staged_ledger/staged_ledger.ml
Comment (mrmr1993) [inline src/lib/staged_ledger/staged_ledger.ml:1036]: Why `#zkApp` here? Would be clearer to have e.g. `number of zkApps in block`.

Comment (mrmr1993) [inline src/lib/staged_ledger/staged_ledger.ml:1036]: I think this log is for profiling. It's unlikely that this will significantly impact timings, so it may make sense to remove this log.

File: src/lib/staged_ledger/staged_ledger.ml
Comment (mrmr1993) [reply src/lib/staged_ledger/staged_ledger.ml:2379]: It would be good to add a test exercising this

File: src/lib/staged_ledger/staged_ledger.ml
Comment (emberian) [reply src/lib/staged_ledger/staged_ledger.ml:1036]: agree, and otherwise it gets a name like `ZkApp_bound_check`

**Comment by ejMina226** [comment]
!ci-build-me

**Comment by ejMina226** [comment]
!ci-build-me

**Comment by ejMina226** [comment]
!ci-build-me

**Comment by ejMina226** [comment]
I did one simple test on this locally. I just ran a single bp node and set the limit to 0. I was told the `apply_diff` method is applied by a bp to itself after the block candidate is produced so when I fired a zkapp transaction at it the block height didn't change. 

I will try a more sophisticated set up next.

**Review by mrmr1993 — APPROVED** [review]

**Comment by mrmr1993** [comment]
!approved-for-mainnet

**Comment by ejMina226** [comment]
!ci-build-me

**Comment by ejMina226** [comment]
I conducted the test as outlined in the RFC (i.e. a network with a seed bp node without this fix and a non-seed/non-bp node with this fix, and checked that the block height never increased for the latter when zkapp transactions started). The test performed as expected.

**Comment by ejMina226** [comment]
I tried the more sophisticated test (as outlined in the ReadMe) and it performed as expected. The node with a lower limit refused to accept a block from another node with a higher-limit.

File: src/lib/block_producer/block_producer.ml
Comment (georgeee) [inline src/lib/block_producer/block_producer.ml:169]: I don't like that two closely related parameters (`zkapp_cmd_limit` and `zkapps_per_block_hard_cap` are so distinctly different in their names. Suggest to name the new parameter as `zkapps_cmd_limit_hardcap` or something all over the codebase

**Review by georgeee — CHANGES_REQUESTED** [review]

File: src/lib/staged_ledger/staged_ledger.ml
Comment (ejMina226) [reply src/lib/staged_ledger/staged_ledger.ml:1050]: Thanks. @georgeee Are you saying there is a function called `is_zkapp` already in the codebase I can use? Or I should rename the function I'm using? Thanks.

Comment (georgeee) [reply src/lib/staged_ledger/staged_ledger.ml:1050]: I was talking of renaming it

Comment (georgeee) [reply src/lib/staged_ledger/staged_ledger.ml:1050]: Just because `is_zkapp` describes well what the function does and is much shorter than the name used currently

Comment (ejMina226) [reply src/lib/staged_ledger/staged_ledger.ml:1050]: Nice. Thanks.

**Comment by ejMina226** [comment]
!ci-build-me

**Comment by ejMina226** [comment]
@georgeee Thanks for the review. Could you approve when you're happy?

**Comment by ejMina226** [comment]
!ci-build-me

**Comment by ejMina226** [comment]
@georgeee Do you know if there is an issue with Nix? The pipeline seems to be failing there. Thanks.

**Comment by georgeee** [comment]
!ci-build-me

**Comment by georgeee** [comment]
!ci-build-me

**Comment by ejMina226** [comment]
!ci-build-me

**Review by georgeee — APPROVED** [review]

**Comment by georgeee** [comment]
@ejMina226 some errors appear on Ci after merge of `berkeley`, could you take a look?

**Comment by ejMina226** [comment]
Hi @georgeee . It's a strange error. It's saying that there's an argument missing to the following function (Zkapp_command.Valid.to_valid) However, I didn't touch this file.

**Comment by ejMina226** [comment]
Actually, I've fixed one issue.

**Comment by ejMina226** [comment]
!ci-build-me

**Comment by mrmr1993** [comment]
!approved-for-mainnet

**Comment by ejMina226** [comment]
@georgeee @mrmr1993 @deepthiskumar Can this be merged?

**Pull request merged**
